### ISSUE TITLE
do_list_unique() - more correct SQL queries

### ISSUE THREAD

**Issue by ghost** [textpattern/textpattern#196]
_From [zasl...@gmail.com](https://code.google.com/u/112308807792196922428/) on May 22, 2012 12:11:05_

What steps will reproduce the problem?

    <txp:article_custom section="ss,ss,ss,ss,ss" category="cc,cc,cc,c2,c2" author="aa,aa,aa,aa" id="1,2,3,4,3,3,3" keywords="kk,kk,kk,kk" />

Duplicate values ​​- is the output of other tags.

What is the expected output?

    select *, unix_timestamp(Posted) as uPosted, unix_timestamp(Expires) as uExpires, unix_timestamp(LastMod) as uLastMod from txp_textpattern as textpattern where 1=1 and Status = 4 and Posted <= now() and (now() <= Expires or Expires = '0000-00-00 00:00:00') and ID IN (1,2,3,4) and (Category1 IN ('cc','c2') or Category2 IN ('cc','c2')) and Section IN ('ss') and AuthorID IN ('aa') and (FIND_IN_SET('kk',Keywords)) order by field(id, 1,2,3,4) limit 0, 10

What do you see instead?

    select *, unix_timestamp(Posted) as uPosted, unix_timestamp(Expires) as uExpires, unix_timestamp(LastMod) as uLastMod from txp_textpattern as textpattern where 1=1 and Status >= 4 and Posted <= now() and (now() <= Expires or Expires = '0000-00-00 00:00:00') and ID IN (1,2,3,4,3,3,3) and (Category1 IN ('cc','cc','cc','c2','c2') or Category2 IN ('cc','cc','cc','c2','c2')) and Section IN ('ss','ss','ss','ss','ss') and AuthorID IN ('aa','aa','aa','aa') and (FIND_IN_SET('kk',Keywords) or FIND_IN_SET('kk',Keywords) or FIND_IN_SET('kk',Keywords) or FIND_IN_SET('kk',Keywords)) order by Posted desc limit 0, 10

Solution:

Replace the call to do_list() do_list_unique()

    function do_list_unique($list, $delim = ','){
        return array_unique(do_list($list, $delim));
    }

This was tested in aks_article plugin.

_Original issue: http://code.google.com/p/textpattern/issues/detail?id=192_

**Comment by makss** [comment]
It's been three years, perhaps the time has come to apply this patch.  :-)

If the objection is not, then I will prepare and test the patch, as described in the first post.
1. Separate function do_list_unique()
function do_list_unique($list, $delim = ',')
{
    return array_unique(array_map('trim', explode($delim, $list)));
}

2. Replace the most critical calls to do_list() do_list_unique()

**Comment by Bloke** [comment]
I'd rather it was a third argument to do_list(), but appreciate the extra conditional inside the function might slow things down a fraction, so maybe your proposed `do_list_unique()` function is better. Given the number of times `do_list()` is called on a typical page, we need to be sensitive to the speed implications.

I don't mind either way, but agree it's a good idea to implement. By all means benchmark them both and make a decision. If using a third argument slower or incurs intensive memory or stack usage in a typical page request, then go with the new function, otherwise, go with the third argument to `do_list()`.

In fact, if we do go with the third argument, how about it becomes something generic like `$flags`? That then allows us to expand the functionality in future by configuring exactly _how_ it works. While I love `do_list()` and use it a lot in my plugins, one thing that repeatedly frustrates me is that it often generates:

    array(
        '0' => '',
    )

So when I come to see if there was anything in the item I just split, I can't simply check if it's empty or not. I need to actually step into the result and test the first argument to see if it's empty. Having some flag to control whether a single, empty result is returned as an actual empty array instead of the "emptyish" structure above would be very handy for me.

Look at `join_atts()` in the 4.6-dev branch for an example of using flags in a function.

Thanks for looking at this. I agree, long overdue.

**Comment by makss** [comment]
I also love the function `do_list()` and often use it in my code. Empty strings do not like me too. :)

Another argument in favor of `do_list_unique()` vs `do_list()` third argument - it's more readable code. Most often used function call with one parameter and it is necessary to indicate clearly each time the second argument:
`do_list_unique($some)` vs `do_list($some, ',', SKIP_DUPE & SKIP_EMPTY_STRING)`

Test Example:

```php
// Proposed new filter constants for variables: STRIP_*

define('STRIP_NONE', 0);

// strip empty string, 0, false, null, empty array(), etc
// as defined http://php.net/manual/en/language.types.boolean.php#language.types.boolean.casting
define('STRIP_EMPTY', 0x1);

// strip only empty string
define('STRIP_EMPTY_STRING', 0x2);

$q = "  www , ee,ee,,,1, 0 ";

print_r( do_list($q) );
print_r( do_list_unique($q) );
print_r( do_list_unique($q, ',', STRIP_EMPTY) );
print_r( do_list_unique($q, ',', STRIP_NONE) );

function do_list_unique($list, $delim = ',', $flags = STRIP_EMPTY_STRING){
	$out = array_unique(array_map('trim', explode($delim, $list)));
	if ($flags & STRIP_EMPTY) {
		$out = array_filter( $out );
	}
	if ($flags & STRIP_EMPTY_STRING) {
		$out = array_filter( $out, function($v){return ($v=='') ? false : true;} );
	}
	return $out;
}

// actual code do_list() from 4.6-dev branch
function do_list($list, $delim = ','){
	return array_map('trim', explode($delim, $list));
}
```

Output:
```
Array			// do_list($q)
(
    [0] => www
    [1] => ee
    [2] => ee
    [3] => 
    [4] => 
    [5] => 1
    [6] => 0
)
Array			// do_list_unique($q) - default strip only empty strings
(
    [0] => www
    [1] => ee
    [5] => 1
    [6] => 0
)
Array			// do_list_unique($q, ',', STRIP_EMPTY)
(
    [0] => www
    [1] => ee
    [5] => 1
)
Array			// do_list_unique($q, ',', STRIP_NONE) - unique
(
    [0] => www
    [1] => ee
    [3] => 
    [5] => 1
    [6] => 0
)
```

**Comment by Bloke** [comment]
I like that! You're right, keeping `do_list()` as it is now is a good idea. `do_list_unique()` has sensible defaults and the various stripping flags make more sense there than they do in the current `do_list()` function.

How I wish PHP supported named args so we didn't have to specify the delimiter just to get access to the flags... *sigh*.

**Comment by Bloke** [comment]
Closed by 52ed2db01ad7f450f1d4339a393ccbfcf706567e
